Core/HW/StereoResampler.cpp

```diff
@@ -0,0 +1,179 @@
+		s16 l2 = m_buffer[indexR2 & INDEX_MASK]; //next
+		s16 r2 = m_buffer[(indexR2 + 1) & INDEX_MASK]; //next
+		int sampleL = ((l1 << 16) + (l2 - l1) * (u16)m_frac) >> 16;
+		int sampleR = ((r1 << 16) + (r2 - r1) * (u16)m_frac) >> 16;
+		samples[currentSample] = clamp_s16(sampleL);  // Do we even need to clamp after interpolation?
+		samples[currentSample + 1] = clamp_s16(sampleR);
+		m_frac += ratio;
+		indexR += 2 * (u16)(m_frac >> 16);
+		m_frac &= 0xffff;
+	}
+
+	int realSamples = currentSample;
+
+	// Padding with the last value to reduce clicking
+	short s[2];
+	s[0] = clamp_s16(m_buffer[(indexR - 1) & INDEX_MASK]);
+	s[1] = clamp_s16(m_buffer[(indexR - 2) & INDEX_MASK]);
+	for (; currentSample < numSamples * 2; currentSample += 2) {
+		samples[currentSample] = s[0];
+		samples[currentSample + 1] = s[1];
+	}
+
+	// Flush cached variable
+	Common::AtomicStore(m_indexR, indexR);
+
+	//if (realSamples != numSamples * 2) {
+	//	ILOG("Underrun! %i / %i", realSamples / 2, numSamples);
+	//}
+
+	return realSamples / 2;
+}
+
+unsigned int StereoResampler::Mix(short* samples, unsigned int num_samples, bool consider_framelimit, int sample_rate) {
+	if (!samples)
+		return 0;
+
+	return m_dma_mixer.Mix(samples, num_samples, consider_framelimit, sample_rate);
+}
+
+void StereoResampler::MixerFifo::PushSamples(const s32 *samples, unsigned int num_samples) {
+	// Cache access in non-volatile variable
+	// indexR isn't allowed to cache in the audio throttling loop as it
+	// needs to get updates to not deadlock.
+	u32 indexW = Common::AtomicLoad(m_indexW);
+
+	// Check if we have enough free space
+	// indexW == m_indexR results in empty buffer, so indexR must always be smaller than indexW
+	if (num_samples * 2 + ((indexW - Common::AtomicLoad(m_indexR)) & INDEX_MASK) >= MAX_SAMPLES * 2)
+		return;
+
+	// AyuanX: Actual re-sampling work has been moved to sound thread
+	// to alleviate the workload on main thread
+	// and we simply store raw data here to make fast mem copy
+	int over_bytes = num_samples * 4 - (MAX_SAMPLES * 2 - (indexW & INDEX_MASK)) * sizeof(short);
+	if (over_bytes > 0) {
+		ClampBufferToS16(&m_buffer[indexW & INDEX_MASK], samples, (num_samples * 4 - over_bytes) / 2);
+		ClampBufferToS16(&m_buffer[0], samples + (num_samples * 4 - over_bytes) / sizeof(short), over_bytes / 2);
+	} else {
+		ClampBufferToS16(&m_buffer[indexW & INDEX_MASK], samples, num_samples * 2);
+	}
+
+	Common::AtomicAdd(m_indexW, num_samples * 2);
+
+	return;
+}
+
+void StereoResampler::PushSamples(const int *samples, unsigned int num_samples) {
+	m_dma_mixer.PushSamples(samples, num_samples);
+}
+
+void StereoResampler::SetDMAInputSampleRate(unsigned int rate) {
+	m_dma_mixer.SetInputSampleRate(rate);
+}
+
+void StereoResampler::MixerFifo::SetInputSampleRate(unsigned int rate) {
+	m_input_sample_rate = rate;
+}
+
+void StereoResampler::DoState(PointerWrap &p) {
+	auto s = p.Section("resampler", 1);
+	if (!s)
+		return;
```

Comment:
I know I'm the crazy savestate guy, but I think it's nice when the savestate resumes things as quickly as possible.

Seems like it'd be easy to dump the audio buffer and parameters, just to be safe and make the audio resume immediately (in case we toggle input sample rate in HLE, I remember seeing something that made it seem like that might happen though I think we normalize first...)  But, I'm probably being overly savestatey here...

-[Unknown]